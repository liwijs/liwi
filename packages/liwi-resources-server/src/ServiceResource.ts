import type { ServiceInterface } from 'liwi-resources';

export interface SubscribeHook<LoggedInUser = unknown, P = unknown> {
  subscribed: (loggedInUser: undefined | LoggedInUser, params: P) => void;
  unsubscribed: (loggedInUser: undefined | LoggedInUser, params: P) => void;
}

export interface ServiceResource<
  ClientService extends ServiceInterface<
    ClientService['queries'],
    ClientService['operations']
  >,
  LoggedInUser = unknown,
> {
  queries: {
    [P in keyof ClientService['queries']]: (
      params: Parameters<ClientService['queries'][P]>[0],
      loggedInUser: undefined | LoggedInUser,
    ) =>
      | ReturnType<ClientService['queries'][P]>
      | Promise<ReturnType<ClientService['queries'][P]>>;
  };
  subscribeHooks?: {
    [P in keyof ClientService['queries']]?: SubscribeHook<
      LoggedInUser,
      Parameters<ClientService['queries'][P]>[0]
    >;
  };
  operations: {
    [P in keyof ClientService['operations']]: (
      params: Parameters<ClientService['operations'][P]>[0],
      loggedInUser: undefined | LoggedInUser,
    ) => ReturnType<ClientService['operations'][P]>;
  };
}
