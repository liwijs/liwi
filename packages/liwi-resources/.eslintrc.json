{
  "parser": "eslint-plugin-typescript/parser",
  "plugins": ["typescript"],
  "extends": ["pob/babel-node", "pob/typescript"],
  "rules": {
    "import/no-cycle": "off"
  }
}
