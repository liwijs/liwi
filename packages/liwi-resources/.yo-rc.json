{
  "pob": {
    "promptValues": {
      "authorName": "Christophe Hurpeau",
      "authorEmail": "christophe@hurpeau.com",
      "authorUrl": "https://christophe.hurpeau.com"
    },
    "pob-config": {
      "documentation": false,
      "testing": false
    }
  },
  "liwi-resources": {
    "type": {
      "type": "lib"
    }
  }
}
