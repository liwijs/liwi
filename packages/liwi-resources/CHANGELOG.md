# Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

## [8.0.3](https://github.com/liwijs/liwi/compare/v8.0.2...v8.0.3) (2021-01-18)

**Note:** Version bump only for package liwi-resources





## [8.0.2](https://github.com/liwijs/liwi/compare/v8.0.1...v8.0.2) (2021-01-18)


### Bug Fixes

* bring back support for webpack 4 ([f2b5583](https://github.com/liwijs/liwi/commit/f2b5583ebfb21c66673b269ed22a9d4a9ffc126d))





## [8.0.1](https://github.com/liwijs/liwi/compare/v8.0.0...v8.0.1) (2021-01-10)

**Note:** Version bump only for package liwi-resources





# [8.0.0](https://github.com/liwijs/liwi/compare/v7.6.2...v8.0.0) (2021-01-10)


### Code Refactoring

* update dev dependencies ([#12](https://github.com/liwijs/liwi/issues/12)) ([c75bfdc](https://github.com/liwijs/liwi/commit/c75bfdcbe5404f5e09679a336edf4bf12b95c57a))


### BREAKING CHANGES

* drop node 10





## [7.2.3](https://github.com/liwijs/liwi/compare/v7.2.2...v7.2.3) (2020-08-08)


### Bug Fixes

* better type ServiceInterface ([34b57b1](https://github.com/liwijs/liwi/commit/34b57b1))





## [7.2.1](https://github.com/liwijs/liwi/compare/v7.2.0...v7.2.1) (2020-08-08)


### Bug Fixes

* params type ([bf54f7b](https://github.com/liwijs/liwi/commit/bf54f7b))





# [7.0.0](https://github.com/liwijs/liwi/compare/v0.18.8...v7.0.0) (2020-08-08)


### Bug Fixes

* allow QueryDescription to have void param ([0c9a894](https://github.com/liwijs/liwi/commit/0c9a894))
* fixes ([482c388](https://github.com/liwijs/liwi/commit/482c388))
* improve ServiceResource definitions ([92a1c3b](https://github.com/liwijs/liwi/commit/92a1c3b))


### Features

* add transformer ([76861ae](https://github.com/liwijs/liwi/commit/76861ae))
* big refactor ([#5](https://github.com/liwijs/liwi/issues/5)) ([a4629c4](https://github.com/liwijs/liwi/commit/a4629c4))
* drop node 8 ([2fb6528](https://github.com/liwijs/liwi/commit/2fb6528))
* queries as fn with result ([1a1abcb](https://github.com/liwijs/liwi/commit/1a1abcb))
* resource operations ([9ddd3b7](https://github.com/liwijs/liwi/commit/9ddd3b7))
* split liwi-resources into server and client ([df73bf9](https://github.com/liwijs/liwi/commit/df73bf9))
* update deps and improve typescript dev ([b42eefb](https://github.com/liwijs/liwi/commit/b42eefb))


### BREAKING CHANGES

* drop node 8
* added Transformer, query model is now value in query description
* multiple typescript typings change
