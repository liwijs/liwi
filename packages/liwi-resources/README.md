<h3 align="center">
  liwi-resources
</h3>

<p align="center">
  resources for liwi
</p>

<p align="center">
  <a href="https://npmjs.org/package/liwi-resources"><img src="https://img.shields.io/npm/v/liwi-resources.svg?style=flat-square"></a>
  <a href="https://npmjs.org/package/liwi-resources"><img src="https://img.shields.io/npm/dw/liwi-resources.svg?style=flat-square"></a>
  <a href="https://npmjs.org/package/liwi-resources"><img src="https://img.shields.io/node/v/liwi-resources.svg?style=flat-square"></a>
  <a href="https://npmjs.org/package/liwi-resources"><img src="https://img.shields.io/npm/types/liwi-resources.svg?style=flat-square"></a>
</p>

## Install

```bash
npm install --save liwi-resources
```

## Usage

```js
import liwiResources from 'liwi-resources';

console.log(liwiResources);
```
