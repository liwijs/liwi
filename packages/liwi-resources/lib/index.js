'use strict';

throw new Error('use liwi-resources-client or liwi-resources-server');
