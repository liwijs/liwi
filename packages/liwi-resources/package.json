{
  "name": "liwi-resources",
  "version": "12.0.0",
  "description": "resources for liwi",
  "keywords": [],
  "author": "Christophe Hurpeau <christophe@hurpeau.com> (https://christophe.hurpeau.com)",
  "license": "ISC",
  "repository": {
    "type": "git",
    "url": "https://github.com/liwijs/liwi.git",
    "directory": "packages/liwi-resources"
  },
  "homepage": "https://github.com/liwijs/liwi",
  "type": "module",
  "engines": {
    "node": ">=22.14.0"
  },
  "sideEffects": false,
  "main": "./dist/index-node.mjs",
  "types": "./dist/definitions/index.d.ts",
  "module": "./dist/index-browser.es.js",
  "browser": "./dist/index-browser.es.js",
  "exports": {
    "./package.json": "./package.json",
    ".": {
      "types": "./dist/definitions/index.d.ts",
      "node": {
        "import": "./dist/index-node.mjs"
      },
      "browser": {
        "import": "./dist/index-browser.es.js"
      }
    }
  },
  "files": [
    "src",
    "dist"
  ],
  "scripts": {
    "build": "yarn clean:build && rollup --config rollup.config.mjs && yarn run build:definitions",
    "build:definitions": "tsc -p tsconfig.json",
    "clean": "yarn clean:build",
    "clean:build": "pob-babel-clean-out dist",
    "lint": "yarn run lint:eslint",
    "lint:eslint": "yarn ../.. run eslint --quiet packages/liwi-resources",
    "watch": "yarn clean:build && rollup --config rollup.config.mjs --watch"
  },
  "browserslist": {
    "production": [
      "defaults",
      "> 0.2%",
      "not ie < 12",
      "not safari < 10",
      "not ios_saf < 10"
    ]
  },
  "pob": {
    "bundler": "rollup-babel",
    "entries": [
      "index"
    ],
    "envs": [
      {
        "target": "node",
        "version": "22",
        "omitVersionInFileName": true
      },
      {
        "target": "browser"
      }
    ]
  },
  "prettier": "@pob/root/prettier-config",
  "dependencies": {
    "extended-json": "12.0.0",
    "liwi-store": "12.0.0"
  },
  "devDependencies": {
    "@babel/core": "7.28.3",
    "@babel/preset-env": "7.28.3",
    "pob-babel": "43.7.0",
    "typescript": "5.9.2"
  }
}
