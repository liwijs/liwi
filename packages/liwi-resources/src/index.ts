import { ExtendedJsonValue } from 'extended-json';
import { Query, QueryParams } from 'liwi-store';

export type {
  Query,
  QuerySubscription,
  QueryParams,
  QueryResult,
  QueryMeta,
  SubscribeCallback,
} from 'liwi-store';

export interface ResourceSubscribePayload<Options> {
  resourceName: string;
  type: string;
  args: Options;
}

export interface ServiceInterface<
  QueryKeys extends keyof any,
  OperationKeys extends keyof any
> {
  queries: {
    [key in QueryKeys]: <Params extends QueryParams<Params> | undefined>(
      params: Params,
    ) => Query<any, Params>;
  };
  operations: {
    [key in OperationKeys]: (params: any) => Promise<any>;
  };
}

// eslint-disable-next-line @typescript-eslint/consistent-type-definitions
export type AckError = {
  code: string;
  message: string;
};

export type ToClientMessage<T = ExtendedJsonValue> = [
  'ack' | 'subscription', // type
  number, // id
  AckError | null, // error
  T, // result
];

export interface DoPayload {
  resourceName: string;
  operationKey: string;
  params: Record<string, ExtendedJsonValue> | undefined;
}

export interface ToServerQueryPayload {
  resourceName: string;
  key: string;
  params: Record<string, ExtendedJsonValue> | undefined;
}

export interface ToServerSubscribeQueryPayload extends ToServerQueryPayload {
  subscriptionId: number;
}

export interface ToServerSubscribeClose {
  subscriptionId: number;
}

// export type ToServerSubscribeQueryChangePayload = ToServerQueryPayload;

type Message<RequestPayload, ResponsePayload> = [
  RequestPayload,
  ResponsePayload,
];

type SubscribeMessage<
  RequestPayload extends {
    params: Record<string, ExtendedJsonValue> | undefined;
  },
  ResponsePayload,
  SubscribeCallbackMessage
> = [RequestPayload, ResponsePayload, SubscribeCallbackMessage];

export interface ToServerSimpleMessages {
  do: Message<DoPayload, undefined>;
  fetch: Message<ToServerQueryPayload, unknown>;
  // 'subscribe:changePayload': Message<
  //   ToServerSubscribeClose & Record<string, ExtendedJsonValue | undefined>,
  //   undefined
  // >;
  'subscribe:close': Message<ToServerSubscribeClose, undefined>;
}

export interface ToServerSubscribeMessages<
  Params extends Record<keyof Params, ExtendedJsonValue> | undefined = never,
  Result = unknown
> {
  subscribe: SubscribeMessage<ToServerSubscribeQueryPayload, undefined, Result>;
  fetchAndSubscribe: SubscribeMessage<
    ToServerSubscribeQueryPayload,
    undefined,
    Result
  >;
}

export type ToServerMessages = ToServerSimpleMessages &
  ToServerSubscribeMessages<Record<string, ExtendedJsonValue> | undefined>;

export type ToServerMessage = (
  | {
      type: 'do';
      payload: ToServerSimpleMessages['do'][0];
    }
  | {
      type: 'fetch';
      payload: ToServerSimpleMessages['fetch'][0];
    }
  // | {
  //     type: 'subscribe:changePayload';
  //     payload: ToServerSimpleMessages['subscribe:changePayload'][0];
  //   }
  | {
      type: 'subscribe:close';
      payload: ToServerSimpleMessages['subscribe:close'][0];
    }
  | {
      type: 'subscribe';
      payload: ToServerSubscribeMessages['subscribe'][0];
    }
  | {
      type: 'fetchAndSubscribe';
      payload: ToServerSubscribeMessages['fetchAndSubscribe'][0];
    }
) & { id: number };

export class ResourcesServerError extends Error {
  code: string;

  constructor(code: string, message: string) {
    super(message);
    this.name = 'ResourcesServerError';
    this.code = code;
  }
}
