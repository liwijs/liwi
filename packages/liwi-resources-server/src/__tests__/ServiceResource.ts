import ts from 'typescript';

const common = `
import { OperationDescription, QueryDescription } from 'liwi-resources';
import { BaseModel } from 'liwi-store';
import ServiceResource from '../ServiceResource';

interface Post extends BaseModel {
  title: string;
}

export interface PostServiceQueries {
  queryAll: QueryDescription<never, Post>;
  queryDetailedPost: QueryDescription<never, Post>;
  // badQuery: {};
}

export interface PostServiceQueriesWithBadQuery {
  queryAll: QueryDescription<never, Post>;
  queryDetailedPost: QueryDescription<never, Post>;
  badQuery: {};
}

`;
//
// const tsRunDiagnostics = () => {
//   let program = ts.createProgram(fileNames, options);
//   let emitResult = program.emit();
//
//   let allDiagnostics = ts
//     .getPreEmitDiagnostics(program)
//     .concat(emitResult.diagnostics);
// }
//
it('should transpile without error on simple queries', () => {
  expect(
    ts.transpileModule(
      `${common}
export const postsResource1: ServiceResource<PostServiceQueries> = {
  queries: {
    queryAll: (params: void, connectedUser) => {
      return {} as any;
    },
    queryDetailedPost: (params: void, connectedUser) => {
      return {} as any;
    },
  },
  operations: {},
};
`,
      {},
    ),
  ).toMatchSnapshot();
});
//
// it('should throw', () => {
//   expect(
//     ts.transpileModule(
//       `${common}
// export const postsResource1: ServiceResource<PostServiceQueries> = {
//   queries: {
//     queryAll: (params: void, connectedUser) => {
//       return {} as any;
//     },
//     queryDetailedPost: (params: void, connectedUser) => {
//       return {} as any;
//     },
//     unexpectedQuery: () => {},
//   },
//   operations: {},
// };
// `,
//       {},
//     ),
//   ).toMatchSnapshot();
// });
//
// // // @dts-jest:pass:snap
// //
// //
// // // @dts-jest:fail:snap
// // export const postsResource2: ServiceResource<PostServiceQueriesWithBadQuery> = {
// //   queries: {},
// //   operations: {},
// // };
// //
// // export interface PostDeleteOperation {
// //   deviceId: string;
// // }
// // export interface PostsServiceOperations {
// //   delete: OperationDescription<PostDeleteOperation, boolean>;
// // }
// //
// // // export const postsResource2: ServiceResource<
// // //   PostServiceQueries,
// // //   PostsServiceOperations
// // // > = {
// // //   queries: {},
// // //   operations: {},
// // // };
