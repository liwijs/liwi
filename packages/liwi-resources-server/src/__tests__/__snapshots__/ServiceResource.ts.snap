// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should transpile without error on simple queries 1`] = `
{
  "diagnostics": [],
  "outputText": ""use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.postsResource1 = void 0;
exports.postsResource1 = {
    queries: {
        queryAll: function (params, loggedInUser) {
            return {};
        },
        queryDetailedPost: function (params, loggedInUser) {
            return {};
        },
    },
    operations: {},
};
",
  "sourceMapText": undefined,
}
`;
